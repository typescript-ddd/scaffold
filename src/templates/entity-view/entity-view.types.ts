import {
  TemplatePropertiesOption,
  TemplatePropertyValue,
  TemplateStringOption,
} from "../shared/template";

export type EntityViewTemplateValues = {
  entityName: string;
  properties: TemplatePropertyValue[];
};

export type EntityViewTemplateOptions = {
  entityName: TemplateStringOption;
  properties: TemplatePropertiesOption;
};
