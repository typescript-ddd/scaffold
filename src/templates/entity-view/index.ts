export * from "./entity-view.generator";
export * from "./entity-view.template";
export * from "./entity-view.types";
