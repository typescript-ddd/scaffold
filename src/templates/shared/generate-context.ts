import { SourceFile } from "ts-morph";

export type ContextImportDeclaration = { moduleSpecifier: string; namedImports: string[] };

export interface GenerateContext {
    resolveDir(type: "domain" | "infrastructure" | "application" | "presentation", subPath?: string): string;
    fileName: string;
    addImportDeclaration(declaration: ContextImportDeclaration): void;
    getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[];
    getModuleSpecifiers(): string[];
    getNamedImports(moduleSpecifier?: string): string[];    
    writeSourceFileImports: (sourceFile: SourceFile) => void;
    clearImportDeclarations(): void;
    tokenize: (text: string, tokens: { name: string }[]) => string;
}