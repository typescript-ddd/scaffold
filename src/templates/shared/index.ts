export * from "./context-builder";
export * from "./generate-context";