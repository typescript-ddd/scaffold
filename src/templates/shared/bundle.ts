import { GenerateContext, Chunk } from "./generate-context";
import { Template, TemplateOption } from "./template";

export interface Bundle<
  TValues extends { [K in keyof TOptions]: any },
  TOptions extends { [key: string]: TemplateOption }
> extends Omit<Template<TValues, TOptions>, "generate"> {
  generate(value: TValues, context: GenerateContext): Chunk[];
}
