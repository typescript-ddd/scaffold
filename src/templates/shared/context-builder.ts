import { GenerateContext, ContextImportDeclaration } from "./generate-context";
import { arrays } from "../../utils";
import { SourceFile } from "ts-morph";

export class ContextBuilder implements GenerateContext {
  private _fileName: string;
  private _importDeclarations: ContextImportDeclaration[] = [];

  private constructor(
    private rootDir: string = "@/src",
    fileName: string = "example.ts"
  ) {
    this._fileName = fileName;
  }

  resolveDir(
    type: "domain" | "infrastructure" | "application" | "presentation",
    subPath?: string
  ): string {
    switch (type) {
      case "domain":
        return `${this.rootDir}/domain${subPath ? `/${subPath}` : ""}`;
      case "infrastructure":
        return `${this.rootDir}/infrastructure${subPath ? `/${subPath}` : ""}`;
      case "application":
        return `${this.rootDir}/application${subPath ? `/${subPath}` : ""}`;
      case "presentation":
        return `${this.rootDir}/presentation${subPath ? `/${subPath}` : ""}`;
      default:
        throw new Error(`Unknown type ${type} for resolveDir`);
    }
  }

  getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[] {
    const modules = this.getModuleSpecifiers();

    const declarations: ContextImportDeclaration[] = [];

    for (const currentModule of modules) {
      const namedImports = this.getNamedImports(currentModule);
      declarations.push({ moduleSpecifier: currentModule, namedImports });
    }

    if (moduleSpecifier) {
      return declarations.filter((x) => x.moduleSpecifier === moduleSpecifier);
    }

    return declarations;
  }

  addImportDeclaration(declaration: ContextImportDeclaration): void {
    this._importDeclarations.push(declaration);
  }

  getModuleSpecifiers(): string[] {
    return arrays.distinct(
      this._importDeclarations.map((x) => x.moduleSpecifier)
    );
  }

  getNamedImports(moduleSpecifier: string): string[] {
    return arrays.distinct(
      this._importDeclarations
        .filter((x) => x.moduleSpecifier === moduleSpecifier)
        .map((x) => x.namedImports)
        .flat()
    );
  }

  writeSourceFileImports(sourceFile: SourceFile) {
    const importDeclarations = this.getImportDeclarations();
    const moduleSpecifiers = this.getModuleSpecifiers();

    const coreImports = importDeclarations.filter(
      (x) => x.moduleSpecifier === "@typescript-ddd/core"
    );

    if (coreImports) {
      sourceFile.addImportDeclaration({
        moduleSpecifier: `@typescript-ddd/core`,
        namedImports: coreImports.map((x) => x.namedImports).flat(),
      });
    }

    // Non core and root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter(
        (x) => x !== "@typescript-ddd/core" && !x.startsWith(this.rootDir)
      )
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }

    // Root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter((x) => x.startsWith(this.rootDir))
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }
  }

  clearImportDeclarations(): void {
    this._importDeclarations = [];
  }

  tokenize(text: string, tokens: { name: string }[]): string {
    let output = text;
    tokens.forEach((token) => {
      output = output.replace(`{{${token.name}}}`, "${" + token.name + "}");
    });
    return "`" + output + "`";
  }

  get fileName(): string {
    return this._fileName;
  }

  static build(rootDir: string): ContextBuilder {
    return new ContextBuilder(rootDir);
  }
}
