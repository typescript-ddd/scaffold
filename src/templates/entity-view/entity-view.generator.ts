import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";
import { EntityViewTemplateValues } from "./entity-view.types";

export const generateEntityView = (
  values: EntityViewTemplateValues,
  context: GenerateContext
) => {
  const { entityName, properties = [] } = values;
  const entityType = strings.capitalize(entityName);
  const viewType = `${entityType}View`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${viewType} {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("domain"),
    namedImports: [entityType],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(viewType);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(viewType)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: [{ name: strings.lower(entityName), type: entityType }],
  });
  constructorDeclaration.setBodyText((writer) => {
    properties.forEach((property) => {
      if (property.name === "id") {
        writer.writeLine(
          `this.id = ${strings.lower(entityName)}.id${
            property.valueType === "string" ? ".value" : ""
          };`
        );
      } else {
        writer.writeLine(
          `this.${property.name} = ${strings.lower(
            entityName
          )}.get${strings.capitalize(property.name)}();`
        );
      }
    });
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(viewType)}.`,
    tags: [{ tagName: "param", text: "{string} id - The entity id." }],
  });

  properties.forEach((property) => {
    classDeclaration
      .addProperty({
        name: property.name,
        type: property.valueType,
        scope: Scope.Public,
        isReadonly: true,
        hasQuestionToken: property.nullable,
      })
      .addJsDoc({
        description: `The ${strings.natural(
          property.name
        )} of the ${strings.natural(entityName)}.`,        
      });
  });

  classDeclaration
    .addMethod({
      name: "create",
      isStatic: true,
      parameters: [{ name: strings.lower(entityName), type: entityType }],
      returnType: viewType,
    })
    .setBodyText((writer) => {
      writer.writeLine(`return new ${viewType}(${strings.lower(entityName)});`);
    });

  classDeclaration
    .addMethod({
      name: "createMany",
      isStatic: true,
      parameters: [
        { name: strings.lower(entityName), type: `${entityType}[]` },
      ],
      returnType: `${viewType}[]`,
    })
    .setBodyText((writer) => {
      writer.writeLine(
        `return ${strings.lower(entityName)}.map(${strings.lower(
          entityName
        )} => ${viewType}.create(${strings.lower(entityName)}));`
      );
    });

  let output = sourceFile.getFullText();

  return output;
};
