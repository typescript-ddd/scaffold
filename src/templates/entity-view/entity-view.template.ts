import { Chunk, GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateEntityView } from "./entity-view.generator";
import {
  EntityViewTemplateOptions,
  EntityViewTemplateValues,
} from "./entity-view.types";

export class EntityViewTemplate
  implements
    Template<EntityViewTemplateValues, EntityViewTemplateOptions>
{
  readonly name = "Entity View";
  readonly description = "Generates an entity view";
  readonly options: EntityViewTemplateOptions = {
    entityName: {
      name: "entityName",
      label: "Entity Name",
      description: "The name of the entity",
      optionType: "string",
    },
    properties: {
        name: "properties",
        label: "Properties",
        description: "The properties of the entity",
        optionType: "properties",
    }
  };
  readonly defaultValues = {
    entityName: "User",
    properties: []
  };
  generate(values: EntityViewTemplateValues, context: GenerateContext, chunkName?: string): Chunk {
    return generateEntityView(values, context, chunkName);
  }
}
