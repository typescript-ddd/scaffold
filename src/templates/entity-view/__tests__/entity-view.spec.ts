import { GenerateContext, ContextBuilder } from "../../shared";
import { generateEntityView } from "../entity-view";

describe("EntityView", () => {
  let context: GenerateContext;

  beforeEach(() => {
    context = ContextBuilder.build("@src");
  });

  it("Should render the entity view template", () => {
    const output = generateEntityView(
      {
        entityName: "User",
        properties: [
          { name: "id", type: "string", prop: "value" },
          { name: "name", type: "string" },
        ],
      },
      context
    );

    expect(output).toMatchSnapshot();
  });
});
