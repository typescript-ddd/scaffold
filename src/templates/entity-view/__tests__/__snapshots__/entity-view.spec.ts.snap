// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityView Should render the entity view template 1`] = `
"import "@typescript-ddd/core";
import { User } from "@src/domain";

/** Represents a user view. */
export class UserView {
    /**
     * Initializes a new instance of a user view.
     * @param {string} id - The entity id.
     */
    private constructor(user: User) {
        this.id = user.id.value;
        this.name = user.getName();
    }

    /** The id of the user. */
    public readonly id: string;
    /** The name of the user. */
    public readonly name: string;

    static create(user: User): UserView {
        return new UserView(user);
    }

    static createMany(user: User[]): UserView[] {
        return user.map(user => UserView.create(user));
    }
}
"
`;
